akka-http-core/src/main/scala/akka/http/impl/engine/client/ProxyGraphStage.scala

```diff
@@ -0,0 +1,119 @@
+import akka.http.scaladsl.settings.ClientConnectionSettings
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, BidiShape, Inlet, Outlet }
+import akka.util.ByteString
+
+object ProxyGraphStage {
+  sealed trait State
+  // Entry state
+  case object Starting extends State
+
+  // State after CONNECT messages has been sent to Proxy and before Proxy responded back
+  case object Connecting extends State
+
+  // State after Proxy responded  back
+  case object Connected extends State
+}
+
+class ProxyGraphStage(targetHostName: String, targetPort: Int, settings: ClientConnectionSettings, log: LoggingAdapter)
+  extends GraphStage[BidiShape[ByteString, ByteString, ByteString, ByteString]] {
+
+  import ProxyGraphStage._
+
+  val bytesIn: Inlet[ByteString] = Inlet("OutgoingTCP.in")
+  val bytesOut: Outlet[ByteString] = Outlet("OutgoingTCP.out")
+
+  val sslIn: Inlet[ByteString] = Inlet("OutgoingSSL.in")
+  val sslOut: Outlet[ByteString] = Outlet("OutgoingSSL.out")
+
+  override def shape: BidiShape[ByteString, ByteString, ByteString, ByteString] = BidiShape.apply(sslIn, bytesOut, bytesIn, sslOut)
+
+  private val connectMsg = ByteString(s"CONNECT ${targetHostName}:${targetPort} HTTP/1.1\r\nHost: ${targetHostName}\r\n\r\n")
+
+  @scala.throws[Exception](classOf[Exception])
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
+    private var state: State = Starting
+
+    val parser = new HttpResponseParser(settings.parserSettings, HttpHeaderParser(settings.parserSettings, log)()) {
+      var first = true
+      override def handleInformationalResponses = false
+      override protected def parseMessage(input: ByteString, offset: Int): StateResult = {
+        if (first) {
+          first = false
+          super.parseMessage(input, offset)
+        } else {
+          emit(RemainingBytes(input.drop(offset)))
+          terminate()
+        }
+      }
+    }
+    parser.setContextForNextResponse(HttpResponseParser.ResponseContext(HttpMethods.CONNECT, None))
+
+    setHandler(sslIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting   ⇒
+          // should not happen
+          case Connecting ⇒
+          // should not happen
+          case Connected ⇒
+            push(bytesOut, grab(sslIn))
+        }
+      }
+    })
+
+    setHandler(bytesIn, new InHandler {
+      override def onPush() = {
+        state match {
+          case Starting ⇒
+          // should not happen to be in Starting state
+          case Connecting ⇒
+            val proxyResponse = grab(bytesIn)
+            parser.parseBytes(proxyResponse) match {
+              case NeedMoreData ⇒ pull(bytesIn)
+              case ResponseStart(StatusCodes.OK, _, _, _, _) ⇒
+                state = Connected
+                pull(sslIn)
+                pull(bytesIn)
```

Comment:
This will probably work because the proxy won't be sending more than the proxy response but in general, if the actual server behind the proxy sends more bytes, the parser might have consumed them. Therefore, just to be sure we should also add that bit of code to make sure that we see all data:

```scala
                    val parseResult = parser.onPull()
                    require(parseResult == ParserOutput.MessageEnd, s"parseResult should be MessageEnd but was $parseResult")
                    parser.onPull() match {
                      case NeedMoreData          ⇒ pull(in) // parser is basically done for our purpose
                      case RemainingBytes(bytes) ⇒ push(out, bytes) // parser already read more than expected, forward that data directly
                      case other ⇒
                        throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
                    }
```